plan9/fs/fs.go

```diff
@@ -0,0 +1,252 @@
+
+	dirsRead []*plan9.Dir // directories read ahead by dirread
+}
+
+func (d *dir9p) ReadDir(n int) ([]fs.DirEntry, error) {
+	var dirs []*plan9.Dir
+	var err error
+
+	if n <= 0 {
+		dirs, err = d.fid.Dirreadall()
+		dirs = append(d.dirsRead, dirs...) // preserve directory order
+		d.dirsRead = d.dirsRead[:0]
+	} else {
+		for err == nil && len(d.dirsRead) < n {
+			dirs, err = d.fid.Dirread()
+			d.dirsRead = append(d.dirsRead, dirs...) // preserve directory order
+		}
+		if len(d.dirsRead) >= n {
+			dirs = d.dirsRead[0:n]
+			d.dirsRead = d.dirsRead[n:]
+		} else {
+			dirs = d.dirsRead
+			d.dirsRead = d.dirsRead[:0]
+		}
+	}
+
+	if len(d.dirsRead) > 0 && err == io.EOF {
+		err = nil
+	}
+
+	entries := make([]fs.DirEntry, len(dirs))
+	for i, dir := range dirs {
+		entries[i] = dirEntry9p{sys: dir}
+	}
+	return entries, err
+}
+
+// fileInfo9p implements io/fs.FileInfo.
+type fileInfo9p struct {
+	sys *plan9.Dir
+}
+
+func (f fileInfo9p) Name() string {
+	return f.sys.Name
+}
+
+func (f fileInfo9p) Size() int64 {
+	// for directories size is implementation defined. Use 0 for portability.
+	if f.sys.Mode&plan9.DMDIR > 0 {
+		return 0
+	}
+
+	// 9p uses uint64 but FileInfo uses int64. For most practical cases,
+	// the conversion it OK.
+	return int64(f.sys.Length)
+}
+
+func (f fileInfo9p) Mode() fs.FileMode {
+	// init mode to the permission bits and then set the others
+	mode := fs.FileMode(f.sys.Mode & 0777)
+	if f.sys.Mode&plan9.DMDIR > 0 {
+		mode |= fs.ModeDir
+	}
+	if f.sys.Mode&plan9.DMAPPEND > 0 {
+		mode |= fs.ModeAppend
+	}
+	if f.sys.Mode&plan9.DMEXCL > 0 {
+		mode |= fs.ModeExclusive
+	}
+	if f.sys.Mode&plan9.DMTMP > 0 {
+		mode |= fs.ModeTemporary
+	}
+
+	// The following are not defined by 9p (http://9p.io/sys/man/5/INDEX.html)
+	// but are defined by the plan9 client
+	if f.sys.Mode&plan9.DMSYMLINK > 0 {
+		mode |= fs.ModeSymlink
+	}
+	if f.sys.Mode&plan9.DMDEVICE > 0 {
+		mode |= fs.ModeDevice
+	}
+	if f.sys.Mode&plan9.DMNAMEDPIPE > 0 {
+		mode |= fs.ModeNamedPipe
+	}
+	if f.sys.Mode&plan9.DMSOCKET > 0 {
+		mode |= fs.ModeSocket
+	}
+	if f.sys.Mode&plan9.DMSETUID > 0 {
+		mode |= fs.ModeSetuid
+	}
+	if f.sys.Mode&plan9.DMSETGID > 0 {
+		mode |= fs.ModeSetgid
+	}
+
+	return mode
+}
+
+func (f fileInfo9p) ModTime() time.Time {
+	return time.Unix(int64(f.sys.Mtime), 0)
+}
+
+func (f fileInfo9p) IsDir() bool {
+	return f.sys.Mode&plan9.DMDIR > 0
+}
+
+func (f fileInfo9p) Sys() interface{} {
+	return f.sys
+}
+
+// dirEntry9p implements io/fs.DirEntry.
+type dirEntry9p struct {
+	sys *plan9.Dir
+}
+
+func (d dirEntry9p) Name() string {
+	return d.sys.Name
+}
+
+func (d dirEntry9p) IsDir() bool {
+	return d.sys.Mode&plan9.DMDIR > 0
+}
+
+func (d dirEntry9p) Info() (fs.FileInfo, error) {
+	return fileInfo9p{sys: d.sys}, nil
+}
+
+func (d dirEntry9p) Type() fs.FileMode {
+	var mode fs.FileMode
+	if d.sys.Mode&plan9.DMDIR > 0 {
+		mode |= fs.ModeDir
+	}
+
+	// The following are not defined by 9p (http://9p.io/sys/man/5/INDEX.html)
```

Comment:
These file modes are provided by an informal protocol extension called [9p2000.u](http://ericvh.github.io/9p-rfc/rfc9p2000.u.html). The file modes only need to be checked if use of 9p2000.u was agreed after protocol version negotiation. See version(9p).